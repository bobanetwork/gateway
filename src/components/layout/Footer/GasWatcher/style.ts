import { Typography } from 'components/global'
import styled, { css } from 'styled-components'
import { mobile, tablet } from 'themes/screens'

export const GasListContainer = styled.div`
  self-align: flex-end;
  display: flex;
  gap: 24px;
  justify-content: flex-end;
  ${tablet(css`
    margin: 10px auto;
  `)}
  ${mobile(css`
    margin: 10px auto;
    display: grid;
    grid-template-columns: repeat(2, 1fr);
  `)}
`

export const GasListItem = styled.div`
  display: flex;
  align-items: center;
  gap: 5px;
`

export const GasListItemLabel = styled(Typography).attrs({
  variant: 'body2',
})`
  color: ${({ theme }) =>
    theme.name === 'light'
      ? theme.colors['gray'][600]
      : theme.colors['gray'][100]};
};
`

export const GasListItemValue = styled(Typography).attrs({
  variant: 'body2',
})`
  color: ${({ theme }) =>
    theme.name === 'light'
      ? theme.colors['gray'][700]
      : theme.colors['gray'][200]};};
`
