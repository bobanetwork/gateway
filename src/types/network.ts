export type NetworkSwitchState =
  | 'idle'
  | 'confirming'
  | 'switching'
  | 'success'
  | 'failed'
  | 'chain-not-found'

export interface NetworkSwitchError {
  code?: number;
  message: string;
}
