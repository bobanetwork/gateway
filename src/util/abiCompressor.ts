/**
 * Script to prepare the human readable abi from the json.
 * based on the functions used out of the abi in network service.
 *
 * to prepare abi execute like
 * ```sh
 *  node ./util/abiCompressor.ts
 * ```
 *
 *
 */

import { ethers } from 'ethers'
import * as fs from 'fs'

const smart_contract_abi_paths = [
  './node_modules/@bobanetwork/core_contracts/artifacts/contracts/L2/messaging/L2StandardBridge.sol/L2StandardBridge.json',
  './node_modules/@bobanetwork/core_contracts/artifacts/contracts/standards/L2StandardERC20.sol/L2StandardERC20.json',
  './node_modules/@bobanetwork/core_contracts/artifacts/contracts/L2/predeploys/OVM_GasPriceOracle.sol/OVM_GasPriceOracle.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/DiscretionaryExitFee.sol/DiscretionaryExitFee.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/LP/L1LiquidityPool.sol/L1LiquidityPool.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/Teleportation.sol/Teleportation.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/LP/L2LiquidityPool.sol/L2LiquidityPool.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/BobaFixedSavings.sol/BobaFixedSavings.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/DAO/governance-token/BOBA.sol/BOBA.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/DAO/governance/GovernorBravoDelegate.sol/GovernorBravoDelegate.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/DAO/governance/GovernorBravoDelegator.sol/GovernorBravoDelegator.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/L2BillingContract.sol/L2BillingContract.json',
  './src/deployment/contracts/L1ERC20.json',
  './src/deployment/contracts/NFTMonsterV2.json',
  './src/deployment/contracts/AuthenticatedFaucet.json',
  './src/deployment/contracts/WAGMIv0.json',
  './src/deployment/contracts/WAGMIv1.json',
  './src/deployment/contracts/crosschain/EthBridge.json',
  './src/deployment/contracts/crosschain/L2StandardERC20.json',
  './src/deployment/contracts/crosschain/LZEndpointMock.json',
]

const networkServicesStr = fs.readFileSync(
  './src/services/networkService.ts',
  'utf8'
)

const compress = () => {
  for (const path of smart_contract_abi_paths) {
    const rawData = fs.readFileSync(path, 'utf8')
    const abiObject = JSON.parse(rawData)
    const name = abiObject.contractName
    let abi = abiObject.abi
    abi = abi.filter((component) => {
      if (component['type'] === 'function' && component['name']) {
        if (networkServicesStr.includes(component['name'])) {
          return true
        }
        return false
      }
    })
    const abiInterface = new ethers.utils.Interface(abi)
    const humanReadableABI = abiInterface.format(ethers.utils.FormatTypes.full)
    const stringifiedHumanReadableABI = JSON.stringify(humanReadableABI)
    const result = `const ${name}ABI = ${stringifiedHumanReadableABI}\n export default ${name}ABI`
    fs.writeFile(`./src/services/abi/${name}.abi.ts`, result, () => {
      console.log(`wrote ${name}`)
    })
  }
}

;(() => {
  compress()
})()
