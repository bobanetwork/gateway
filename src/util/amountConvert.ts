/*
Copyright 2021-present Boba Network.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

// we use BigNumber here for decimal support
import BigNumber from 'bignumber.js'
import { BigNumberish } from 'ethers'

export const logAmount = (
  amount: BigNumberish,
  power: number,
  truncate = 0
) => {
  // toString as some methods seem to supply BigNumber
  const x = new BigNumber(amount?.toString())
  const exp = new BigNumber(10).pow(power)

  const calculated = x.div(exp)

  if (truncate > 0) {
    return calculated.toFixed(truncate)
  } else {
    return calculated.toFixed()
  }
}

/*Takes a value such as 3.92 and converts it into
a BigNumber in wei

Duplicates

ethers.utils.parseUnits( valueString , decimalsOrUnitName )   =>   BigNumber
*/

export const powAmount = (amount: number, decimals: number) => {
  const x = new BigNumber(amount)
  const exp = new BigNumber(10).pow(decimals)

  const calculated = x.multipliedBy(exp)
  return calculated.toFixed(0)
}

/* more clearly named version of this */
export const toWei_String = (amount: number, decimals: number) => {
  const x = new BigNumber(amount)
  const exp = new BigNumber(10).pow(decimals)

  const calculated = x.multipliedBy(exp)
  return calculated.toFixed(0)
}

const lookupMap = {
  ETH: { provider: 'ethereum' },
  BOBA: { provider: 'boba-network' },
  OLO: { provider: 'oolongswap' },
  OMG: { provider: 'omisego' },
  USDC: { provider: 'usd-coin' },
  BNB: { provider: 'binancecoin' },
  tBNB: { provider: 'binancecoin' },
}

export const amountToUsd = (amount: number, lookupPrice, token) => {
  const { symbol } = token
  if (!symbol) {
    return 0
  }
  const currencyInfo = lookupMap[symbol]
  if (!!currencyInfo && !!lookupPrice[currencyInfo.provider]) {
    return amount * lookupPrice[currencyInfo.provider].usd
  } else if (!!lookupPrice[symbol.toLowerCase()]) {
    return amount * lookupPrice[symbol.toLowerCase()].usd
  } else {
    return 0
  }
}

export const formatLargeNumber = (num: number) => {
  const exp = Math.floor(Math.log10(Math.abs(num)) / 3)
  const scaledNum = Math.abs(num) / Math.pow(10, exp * 3)
  const formattedNumber = scaledNum.toFixed(2)
  const exponent = ['', 'k', 'M', 'B', 'T'][exp] || ''
  return isNaN(Number(formattedNumber)) ? '0' : formattedNumber + exponent
}
