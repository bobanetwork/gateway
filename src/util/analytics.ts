import ReactGA from 'react-ga4'
import { GA4_MEASUREMENT_ID } from './constant'

export const init = () => {
  if (window.location.hostname === 'localhost') {
    return null
  }
  if (GA4_MEASUREMENT_ID) {
    // init if the GA4 Measurement Id is available.
    ReactGA.initialize(GA4_MEASUREMENT_ID?.toString())
  }
}

export const sendPageView = (path: string) => {
  ReactGA.send({
    hitType: 'pageview',
    page: path,
    title: path,
  })
}

export const trackClick = (category: string, action: string, label: string) => {
  ReactGA.event({
    category,
    action,
    label,
  })
}

export default {
  init,
  sendPageView,
  trackClick,
}
