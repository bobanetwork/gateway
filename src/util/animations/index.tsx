import React, { useState, useEffect } from 'react'
import { TConductorInstance } from 'react-canvas-confetti/dist/types'
import Realistic from 'react-canvas-confetti/dist/presets/realistic'
import { TCanvasConfettiAnimationOptions } from 'react-canvas-confetti/dist/types/normalization'
import styled from 'styled-components'
import dayjs from 'dayjs'

const CanvaContainer = styled.div`
  display: block;
  z-index: 9999;
  position: fixed;
`
const count = 200
const defaults = {
  origin: { y: 0.7 },
}

export const Confetti = ({ size, particles, spreads }: any) => {
  const [conductor, setConductor] = useState<TConductorInstance>()

  const onInit = ({ conductor }: { conductor: TConductorInstance }) => {
    setConductor(conductor)
  }

  const onOnce = () => {
    conductor?.shoot()
  }

  const PARTICLES = {
    heart: [
      {
        type: 'path',
        path: 'M167 72c19,-38 37,-56 75,-56 42,0 76,33 76,75 0,76 -76,151 -151,227 -76,-76 -151,-151 -151,-227 0,-42 33,-75 75,-75 38,0 57,18 76,56z',
        matrix: [
          0.03333333333333333, 0, 0, 0.03333333333333333, -5.566666666666666,
          -5.533333333333333,
        ],
      },
    ],
    circle: ['circle'],
    star: ['star'],
  }

  const heartsOptions = (options: TCanvasConfettiAnimationOptions) => {
    return {
      ...options,
      startVelocity: 50,
      decay: 0.95,
      spread: 180,
      zIndex: 9999,
      gravity: 3,
      origin: {
        y: 1,
      },
      scalar: size || 7,
      particleCount: Math.floor(count * (Math.random() * (0.5 - 0.1) + 0.1)),
      colors: ['#AEDB01'],
      shapes: PARTICLES[particles] || ['square'],
    }
  }

  useEffect(() => {
    setTimeout(() => {
      onOnce()
    }, 1500)
  })

  return (
    <CanvaContainer>
      <Realistic onInit={onInit} decorateOptions={heartsOptions} />
    </CanvaContainer>
  )
}

export const ValentinesDayBanner = () => {
  const [isValentinesDay, setIsValentinesDay] = useState(false)

  useEffect(() => {
    const today = dayjs()
    const isFebruary14 = today.date() === 14 && today.month() === 1
    if (isFebruary14) {
      setIsValentinesDay(true)
    }
  }, [])
  return (
    <>
      {isValentinesDay && (
        <>
          <Confetti size={8} particles="heart" spreads={150} />
          <Confetti size={2} particles="heart" spreads={150} />
          <Confetti size={1} particles="heart" spreads={150} />
        </>
      )}
    </>
  )
}
