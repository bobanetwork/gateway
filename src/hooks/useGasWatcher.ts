import {
  fetchVerifierStatus,
  resetVerifierStatus,
} from 'actions/verifierAction'
import { useCallback, useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import {
  selectBaseEnabled,
  selectActiveNetworkName,
  selectActiveNetwork,
  selectActiveNetworkType,
  selectVerifierStatus,
} from 'selectors'
import networkService from 'services/networkService'
import { Network, NetworkType } from 'util/network/network.util'
import { fetchGasDetail } from 'services/gas.service'
import useInterval from './useInterval'
import { GAS_POLL_INTERVAL } from 'util/constant'

/**
 *
 * useGasWatcher.
 *
 * Gas Calculation
 *
   Fetch gas savings (ONLY ETH MAINNET) & gas details.
   The l1 security fee is moved to the l2 fee
   const gasSavings = (Number(gas.gasL1) * (l2Fee - l1SecurityFee) / Number(gas.gasL2)) / l2Fee;
   The l1 security fee is directly deducted from the user's account
 *
 * NOTE:TODO: https://github.com/bobanetwork/boba/pull/982#discussion_r1253868688
 */

const useGasWatcher = () => {
  const dispatch = useDispatch<any>()

  const [gas, setGas] = useState<any>()
  const [savings, setSavings] = useState<number>(1)

  const verifierStatus = useSelector(selectVerifierStatus)
  const baseEnabled = useSelector(selectBaseEnabled())
  const networkName = useSelector(selectActiveNetworkName())
  const activeNetwork = useSelector(selectActiveNetwork())
  const activeNetworkType = useSelector(selectActiveNetworkType())

  const loadGasDetail = useCallback(() => {
    if (baseEnabled) {
      const fetchGas = async () => {
        const gasDetail = await fetchGasDetail()
        setGas(gasDetail)
      }

      fetchGas()
      if (activeNetwork === Network.ETHEREUM) {
        dispatch(fetchVerifierStatus())
      } else {
        dispatch(resetVerifierStatus())
      }
    }
  }, [networkName, baseEnabled, dispatch])

  useEffect(() => {
    const getGasSavings = async () => {
      const l1SecurityFee = await networkService.estimateL1SecurityFee()
      const l2Fee = await networkService.estimateL2Fee()

      const gasSavings: any =
        (Number(gas.gasL1) * l2Fee) /
        Number(gas.gasL2) /
        (l2Fee + l1SecurityFee)

      setSavings(
        gasSavings && gasSavings.toString() !== 'Infinity' ? gasSavings : 0
      )
    }
    // Load gas savings only in case of ETHEREUM MAINNET
    if (
      activeNetwork === Network.ETHEREUM &&
      activeNetworkType === NetworkType.MAINNET &&
      gas
    ) {
      getGasSavings()
    } else {
      setSavings(1)
    }
  }, [gas, activeNetwork])

  useEffect(() => {
    loadGasDetail()
  }, [loadGasDetail, activeNetwork])

  useInterval(() => {
    loadGasDetail()
  }, GAS_POLL_INTERVAL)

  return { savings, gas, verifierStatus }
}

export default useGasWatcher
