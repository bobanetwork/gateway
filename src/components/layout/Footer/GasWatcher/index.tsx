import React, { FC } from 'react'
import {
  GasListContainer,
  GasListItem,
  GasListItemLabel,
  GasListItemValue,
} from './style'
import useGasWatcher from 'hooks/useGasWatcher'
import { useSelector } from 'react-redux'
import { selectActiveNetwork, selectActiveNetworkName } from 'selectors'
import { Network } from 'util/network/network.util'

const GasWatcher: FC = () => {
  const { gas, savings, verifierStatus } = useGasWatcher()
  const networkName = useSelector(selectActiveNetworkName())
  const activeNetwork = useSelector(selectActiveNetwork())

  if (!gas) {
    return null
  }

  return (
    <GasListContainer id="gasDetails">
      <GasListItem>
        <GasListItemLabel>{networkName['l1']}</GasListItemLabel>
        <GasListItemValue>{gas?.gasL1} Gwei</GasListItemValue>
      </GasListItem>
      <GasListItem>
        <GasListItemLabel>{networkName['l2']}</GasListItemLabel>
        <GasListItemValue>{gas?.gasL2} Gwei</GasListItemValue>
      </GasListItem>
      <GasListItem>
        <GasListItemLabel>Savings</GasListItemLabel>
        <GasListItemValue>{savings?.toFixed(0)}x</GasListItemValue>
      </GasListItem>
      <GasListItem>
        <GasListItemLabel>L1</GasListItemLabel>
        <GasListItemValue>{gas?.blockL1}</GasListItemValue>
      </GasListItem>
      <GasListItem>
        <GasListItemLabel>L2</GasListItemLabel>
        <GasListItemValue>{gas?.blockL2}</GasListItemValue>
      </GasListItem>
      {activeNetwork === Network.ETHEREUM ? (
        <GasListItem>
          <GasListItemLabel>Last Verified Block</GasListItemLabel>
          <GasListItemValue>
            {Number(verifierStatus?.matchedBlock || 0)}
          </GasListItemValue>
        </GasListItem>
      ) : null}
    </GasListContainer>
  )
}

export default GasWatcher
