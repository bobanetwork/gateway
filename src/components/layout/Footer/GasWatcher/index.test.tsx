import { render, screen } from '@testing-library/react'
import React from 'react'
import GasWatcher from '.'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import CustomThemeProvider from 'themes'
import configureStore from 'redux-mock-store'
import { Network, NetworkType } from 'util/network/network.util'
import useGasWatcher from 'hooks/useGasWatcher'

const mockStore = configureStore()

jest.mock('hooks/useGasWatcher')

const mockUseGasWatcher = useGasWatcher as jest.MockedFunction<
  typeof useGasWatcher
>

const renderGasWatcher = ({ options = {} }: any) => {
  return render(
    <MemoryRouter>
      <Provider
        store={mockStore({
          ui: {
            theme: 'dark',
          },
          network: {
            activeNetwork: Network.ETHEREUM,
            activeNetworkType: NetworkType.MAINNET,
            activeNetworkName: {
              l1: 'Ethereum',
              l2: 'Boba ETH',
            },
          },
          ...options,
        })}
      >
        <CustomThemeProvider>
          <GasWatcher />
        </CustomThemeProvider>
      </Provider>
    </MemoryRouter>
  )
}

describe('Footer GasWatcher', () => {
  test('should match snapshot when gas value empty', () => {
    mockUseGasWatcher.mockReturnValue({
      gas: null,
      savings: 1,
      verifierStatus: 231223,
    })
    const { asFragment } = renderGasWatcher({})
    expect(asFragment()).toMatchSnapshot()
  })

  test('should match snapshot when gas has value', () => {
    mockUseGasWatcher.mockReturnValue({
      gas: {
        gasL1: 10,
        gasL2: 20,
        blockL1: 110000,
        blockL2: 220000,
      },
      savings: 1,
      verifierStatus: 231223,
    })
    const { asFragment } = renderGasWatcher({})
    expect(asFragment()).toMatchSnapshot()
  })

  test('should show last verified block in case of ethereum network', () => {
    mockUseGasWatcher.mockReturnValue({
      gas: {
        gasL1: 10,
        gasL2: 20,
        blockL1: 110000,
        blockL2: 220000,
      },
      savings: 1,
      verifierStatus: 231223,
    })
    renderGasWatcher({})
    expect(screen.getByText('Last Verified Block')).toBeInTheDocument()
  })

  test('should remove last verified block from dom incase of BNB network', () => {
    mockUseGasWatcher.mockReturnValue({
      gas: {
        gasL1: 10,
        gasL2: 20,
        blockL1: 110000,
        blockL2: 220000,
      },
      savings: 1,
      verifierStatus: 231223,
    })
    renderGasWatcher({
      options: {
        network: {
          activeNetwork: Network.BNB,
          activeNetworkType: NetworkType.MAINNET,
          activeNetworkName: {
            l1: 'Binance',
            l2: 'Boba BNB',
          },
        },
      },
    })
    expect(screen.queryByText('Last Verified Block')).not.toBeInTheDocument()
  })
})
