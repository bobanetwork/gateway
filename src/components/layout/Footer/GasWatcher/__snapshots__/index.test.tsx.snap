// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer GasWatcher should match snapshot when gas has value 1`] = `
<DocumentFragment>
  <div
    class="sc-fPXMVe ktmBcq"
    id="gasDetails"
  >
    <div
      class="sc-gFqAkR emfUQK"
    >
      <p
        class="sc-dAlyuH eGlSty sc-ikkxIA bNNvzA"
      >
        Ethereum
      </p>
      <p
        class="sc-dAlyuH eGlSty sc-dAbbOL nalzD"
      >
        10 Gwei
      </p>
    </div>
    <div
      class="sc-gFqAkR emfUQK"
    >
      <p
        class="sc-dAlyuH eGlSty sc-ikkxIA bNNvzA"
      >
        Boba ETH
      </p>
      <p
        class="sc-dAlyuH eGlSty sc-dAbbOL nalzD"
      >
        20 Gwei
      </p>
    </div>
    <div
      class="sc-gFqAkR emfUQK"
    >
      <p
        class="sc-dAlyuH eGlSty sc-ikkxIA bNNvzA"
      >
        Savings
      </p>
      <p
        class="sc-dAlyuH eGlSty sc-dAbbOL nalzD"
      >
        1x
      </p>
    </div>
    <div
      class="sc-gFqAkR emfUQK"
    >
      <p
        class="sc-dAlyuH eGlSty sc-ikkxIA bNNvzA"
      >
        L1
      </p>
      <p
        class="sc-dAlyuH eGlSty sc-dAbbOL nalzD"
      >
        110000
      </p>
    </div>
    <div
      class="sc-gFqAkR emfUQK"
    >
      <p
        class="sc-dAlyuH eGlSty sc-ikkxIA bNNvzA"
      >
        L2
      </p>
      <p
        class="sc-dAlyuH eGlSty sc-dAbbOL nalzD"
      >
        220000
      </p>
    </div>
    <div
      class="sc-gFqAkR emfUQK"
    >
      <p
        class="sc-dAlyuH eGlSty sc-ikkxIA bNNvzA"
      >
        Last Verified Block
      </p>
      <p
        class="sc-dAlyuH eGlSty sc-dAbbOL nalzD"
      >
        0
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Footer GasWatcher should match snapshot when gas value empty 1`] = `<DocumentFragment />`;
